### PR TITLE
Add Language Detector to Text Classifier

### PR DESCRIPTION
**Pull request by LorenzoMinto**
Resolves https://github.com/brave/brave-browser/issues/16072
Add Language Detector to Text Classifier

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc [hunk 1]
```diff
@@ -31,6 +35,17 @@ std::string GetTopSegmentFromPageProbabilities(
   return iter->first;
 }
 
+chrome_lang_id::NNetLanguageIdentifier::Result detect_page_language(const std::string& text) {
+
+  const std::string contents = text;
+
+  chrome_lang_id::NNetLanguageIdentifier lang_id;
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc:42]: nit: no need to shorten `language_id`

File: vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc [hunk 2]
```diff
@@ -31,6 +35,17 @@ std::string GetTopSegmentFromPageProbabilities(
   return iter->first;
 }
 
+chrome_lang_id::NNetLanguageIdentifier::Result detect_page_language(const std::string& text) {
+
+  const std::string contents = text;
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc:40]: nit: could pass text directly into L44

File: vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc [hunk 3]
```diff
@@ -31,6 +35,17 @@ std::string GetTopSegmentFromPageProbabilities(
   return iter->first;
 }
 
+chrome_lang_id::NNetLanguageIdentifier::Result detect_page_language(const std::string& text) {
+
+  const std::string contents = text;
+
+  chrome_lang_id::NNetLanguageIdentifier lang_id;
+  const chrome_lang_id::NNetLanguageIdentifier::Result lang_id_result =
+      lang_id.FindTopNMostFreqLangs(contents, /*num_langs=*/1).at(0);
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc:44]: nit: in ads/b-c we usually use `lang_id.FindTopNMostFreqLangs(contents, /* num_langs */ 1).at(0);`

File: vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc [hunk 4]
```diff
@@ -48,6 +63,20 @@ void TextClassification::Process(const std::string& text) {
     return;
   }
 
+  const chrome_lang_id::NNetLanguageIdentifier::Result res_id = detect_page_language(text);
+
+  const std::string locale = brave_l10n::LocaleHelper::GetInstance()->GetLocale();
+  const std::string locale_language = brave_l10n::GetLanguageCode(locale);
+
+  // NOTE: This holds as long as the resource is univocally matched to the locale
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc:71]: nit: remove comment

File: vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc [hunk 5]
```diff
@@ -48,6 +63,20 @@ void TextClassification::Process(const std::string& text) {
     return;
   }
 
+  const chrome_lang_id::NNetLanguageIdentifier::Result res_id = detect_page_language(text);
+
+  const std::string locale = brave_l10n::LocaleHelper::GetInstance()->GetLocale();
+  const std::string locale_language = brave_l10n::GetLanguageCode(locale);
+
+  // NOTE: This holds as long as the resource is univocally matched to the locale 
+  if (res_id.is_reliable && locale_language != res_id.language) {
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc:72]: so this means that if we have low confidence in the detection, we classify? Why not `if (!res_id.is_reliable || locale_language != res_id.language)`?

Comment (LorenzoMinto) [reply vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/contextual/text_classification/text_classification_processor.cc:72]: My thinking was: if the model has low confidence then disregard the model and go with default behaviour. But it's true that even unreliability of the detection is some actionable information. If we trust `cld_3` your condition is probably best.

**Comment by bsclifton** [comment]
Closing as stale

**Pull request closed without merging**